x/lightclient/keeper/client.go

```diff
@@ -0,0 +1,71 @@
+package keeper
+
+import (
+	"time"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	ibcclienttypes "github.com/cosmos/ibc-go/v7/modules/core/02-client/types"
+	ibctm "github.com/cosmos/ibc-go/v7/modules/light-clients/07-tendermint"
+	"github.com/dymensionxyz/dymension/v3/x/lightclient/types"
+	rollapptypes "github.com/dymensionxyz/dymension/v3/x/rollapp/types"
+)
+
+// GetProspectiveCanonicalClient returns the client id of the first IBC client which can be set as the canonical client for the given rollapp.
+// Returns empty string if no such client is found.
+func (k Keeper) GetProspectiveCanonicalClient(ctx sdk.Context, rollappId string, stateInfo *rollapptypes.StateInfo) (clientID string) {
+	clients := k.ibcClientKeeper.GetAllGenesisClients(ctx)
+	for _, client := range clients {
+		clientState, err := ibcclienttypes.UnpackClientState(client.ClientState)
+		if err != nil {
+			continue
+		}
+		// Cast client state to tendermint client state - we need this to get the chain id and state height
+		tmClientState, ok := clientState.(*ibctm.ClientState)
+		if !ok {
+			continue
+		}
+		if tmClientState.ChainId != rollappId {
+			continue
+		}
+		if !types.IsCanonicalClientParamsValid(tmClientState) {
+			continue
+		}
+		sequencerPk, err := k.GetSequencerPubKey(ctx, stateInfo.Sequencer)
+		if err != nil {
+			continue
+		}
+		for _, bd := range stateInfo.GetBDs().BD {
+			height := ibcclienttypes.NewHeight(1, bd.GetHeight())
```

Comment:
Here, instead of iterating over BDs in the update, you need to take the latest height of the update
subtract 1, and then compare against ALL client consensus states less than or equal to that